src/core/thread/network_data_service.cpp

```diff
@@ -195,20 +195,68 @@ Error Manager::GetNextDnsSrpAnycastInfo(Iterator &aIterator, DnsSrpAnycast::Info
 
 Error Manager::FindPreferredDnsSrpAnycastInfo(DnsSrpAnycast::Info &aInfo) const
 {
+    Error               error = kErrorNotFound;
     Iterator            iterator;
     DnsSrpAnycast::Info info;
-    bool                found = false;
+    DnsSrpAnycast::Info maxSeqNumInfo;
+
+    // Determine the entry with largest seq number in two ways:
+    // `aInfo` will track the largest using serial number arithmetic
+    // comparison, while `maxSeqNumInfo` tracks the largest using
+    // normal compassion.
 
     while (GetNextDnsSrpAnycastInfo(iterator, info) == kErrorNone)
     {
-        if (!found || info.IsSequenceNumberAheadOf(aInfo))
+        if (error == kErrorNotFound)
+        {
+            aInfo         = info;
+            maxSeqNumInfo = info;
+            error         = kErrorNone;
+            continue;
+        }
+
+        if (SerialNumber::IsGreater(info.mSequenceNumber, aInfo.mSequenceNumber))
         {
             aInfo = info;
-            found = true;
+        }
+
+        if (info.mSequenceNumber > maxSeqNumInfo.mSequenceNumber)
+        {
+            maxSeqNumInfo = info;
+        }
+    }
+
+    SuccessOrExit(error);
+
+    // Check that the largest seq number using serial number arithmetic is
+    // well-defined (i.e., the seq number is larger than all other seq
+    // numbers values). If it is not, we prefer `maxSeqNumInfo` entry.
+
+    iterator.Reset();
+
+    while (GetNextDnsSrpAnycastInfo(iterator, info) == kErrorNone)
+    {
+        constexpr uint8_t kMidValue = (NumericLimits<uint8_t>::kMax / 2) + 1;
+        uint8_t           seqNumber = info.mSequenceNumber;
+        uint8_t           diff;
+
+        if (seqNumber == aInfo.mSequenceNumber)
+        {
+            continue;
+        }
+
+        diff = seqNumber - aInfo.mSequenceNumber;
+
+        if ((diff == kMidValue) || !SerialNumber::IsGreater(aInfo.mSequenceNumber, seqNumber))
```

Comment:
```suggestion
        if (!SerialNumber::IsGreater(aInfo.mSequenceNumber, seqNumber - 1))
```
Is it the same (and correct)?

Reply:
Agree that the current code is more readable. 
Thanks